### DISCUSSION TITLE
💡 Road to Richer UI components in Foam VS Code extension?

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [foambubble/foam#951] [Ideas]
Hi Foam team & community,
I simply wanted to share some thoughts on how we could perhaps create richer web views inside of Foam, if it makes sense and is of interest to others.

Recently Github/MSFT team have been enabling richer APIs and libraries to allow VS Code extension creators the ability to create rich web UI's within VS Code. For example. Microsoft and Github collaborated to create this library ([vscode-webview-ui-toolkit](https://github.com/microsoft/vscode-webview-ui-toolkit)), which allows extension authors to author Webviews with rich Javascript frameworks like React etc.

Video comes from this discussion I started on [Twitter in this thread](https://twitter.com/i/status/1491444304019345409)

https://user-images.githubusercontent.com/6743796/156423151-45287468-c1de-4ff3-b98a-567c64f6fd0b.mp4

One minor thing that is lacking is the ability to create the Foam Notes graph with custom filter buttons and toggles like this:
Credits to this article for the image: _[It’s Not Just a Pretty Gimmick: In Defense of Obsidian’s Graph View](https://www.obsidianroundup.org/its-not-just-a-pretty-gimmick-in-defense-of-obsidians-graph-view/)_

![image](https://user-images.githubusercontent.com/6743796/156423433-bd23eaaf-cbac-42ea-b8bb-6ae423beaea2.png)

## Considerations
1. With the ability to create custom UI's (essentially mini web apps), it not careful, it would be easy to for Foam to grow in complexity and spiral out of focus/control.

## Question(s)
1. Is any of this of interest to folks using Foam?
2. Maybe to address the point 1️⃣ in considerations section, an API from Foam's core code, can be created to others to build or embed their own UI's on top of Foam's API. In the long run, this would allow the Foam project to remain focused on creating the best Zettlekasten experience; focusing on nailing the API internals can only be a good thing 

### References
- [React Webview UI Toolkit for VS Code](https://next.github.com/projects/react-webview-ui-toolkit/)
- [It’s Not Just a Pretty Gimmick: In Defense of Obsidian’s Graph View](https://www.obsidianroundup.org/its-not-just-a-pretty-gimmick-in-defense-of-obsidians-graph-view/)
- [Think Better with the Graph View + Live note-making session](https://youtu.be/KuzBiqesNdo?t=29)
- [Twitter thread](https://twitter.com/i/status/1491444304019345409)

**Comment by riccardoferretti** [comment]
Thanks for the pointer @cliffordfajardo, this is very useful!

Here are my comments: 

> With the ability to create custom UI's (essentially mini web apps), it not careful, it would be easy to for Foam to grow in complexity and spiral out of focus/control.

Agreed, and there has been some talk about what would be considered useful and in scope for Foam wrt to the dataviz graph. 
You can get an idea of various proposals here: https://github.com/foambubble/foam/labels/foam-dataviz

> Is any of this of interest to folks using Foam?

Yes, very much so!
As of today we have been using DAT only because it was an easy way to create some UI controls for the graph. A migration to the VS Code component library feels like a very good path to me.

> Maybe to address the point 1️⃣ in considerations section, an API from Foam's core code, can be created to others to build or embed their own UI's on top of Foam's API. In the long run, this would allow the Foam project to remain focused on creating the best Zettlekasten experience; focusing on nailing the API internals can only be a good thing

Foam is about to publish an API (in alpha at first) that can be used by other extensions to augment the note taking/browsing experience, as well as add additional semantics (e.g. todos, calendars, ...) to Foam, while keeping the core lean, and focused on the parsing of metadata from the knowledge base, therefore exposing the model, and providing building blocks as commands/API.

A few examples:
- https://github.com/foambubble/foam/issues/924
- https://github.com/foambubble/foam/issues/894

---

To sum up, yes, using a native VS Code feel for the graph sounds good.
The first step to me would be migration of the current view to this UI style (and get rid of DAT).
For further development, keep in mind what we want Foam core to include, and what hooks we want to expose to third party extensions built on top of it.
Any help/support/idea is always welcome.
